File: julia/sieve/sieve.jl
function sieve(limit::Integer)
    nonprime = zero(limit)
    sieve = collect(1:limit)
    sieve[1] = nonprime
    i = 2
    while i^2 <= limit
        step = i * (i == 2 ? 1 : 2)
        for j in i^2:step:limit
            sieve[j] = nonprime
        end
        i += 1
    end
    filter(i -> i != nonprime, sieve)
end

# Erik's implementation
#
#   function sieve(limit::Integer)
#       s = fill(true, limit)
#       s[1] = false
#       for i=2:limit j=2i:i:limit
#           s[j] = false
#       end
#        findall(s)
#   end

File: java/sieve/src/main/java/Sieve.java
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Sieve {
    private final int max;
    private final List<Integer> primes;

    Sieve(int maxPrime) {
        max = maxPrime;
        primes = eratosthenes();
    }

    List<Integer> getPrimes() {
        return primes;
    }

    private List<Integer> eratosthenes() {
        List<Integer> numbers = IntStream
                .rangeClosed(0, max)
                .mapToObj(Integer::new)
                .collect(Collectors.toList());

        removeMultiples(numbers, 2);
        
        int sqrt = (int) Math.floor(Math.sqrt(max));
        for (int i = 3; i <= sqrt; i += 2)
            removeMultiples(numbers, i);

        return numbers
                .stream()
                .filter(i -> i != null && i >= 2)
                .collect(Collectors.toList());
    }

    private void removeMultiples(List<Integer> numbers, int n) {
        int step = 2 * (n == 2 ? 1 : n);
        for (int i = n * n; i <= max; i += step) {
            numbers.set(i, null);
        }
    }
}
